services/src/test/java/org/keycloak/vault/VaultTranscriberTest.java

```diff
@@ -0,0 +1,210 @@
+        }
+        // after the try-with-resources block the secret should have been overridden
+        Assert.assertFalse(Arrays.equals("secret2".toCharArray(), secretBuffer.array()));
+
+        // now attempt to obtain a secret using a proper vault string with a non-existing key.
+        try (VaultCharSecret secret = transcriber.getCharSecret("${vault.non_existing_key}")) {
+            Optional<CharBuffer> optional = secret.getCharSecret();
+            Assert.assertFalse(optional.isPresent());
+        }
+
+        // finally try to obtain a secret using a string that is not a vault string. The string itself should be encoded.
+        try (VaultCharSecret secret = transcriber.getCharSecret("non_vault_secret")) {
+            Optional<CharBuffer> optional = secret.getCharSecret();
+            Assert.assertTrue(optional.isPresent());
+            secretBuffer = optional.get();
+            Assert.assertTrue(Arrays.equals("non_vault_secret".toCharArray(), secretBuffer.array()));
+        }
+        // after the try-with-resources block the secret should have been overridden
+        Assert.assertFalse(Arrays.equals("non_vault_secret".toCharArray(), secretBuffer.array()));
+
+        // check that a null expression results in an empty optional instance.
+        try (VaultCharSecret secret = transcriber.getCharSecret(null)) {
+            Assert.assertFalse(secret.getCharSecret().isPresent());
+        }
+    }
+
+    @Test
+    public void testRetrieveStringSecret() throws Exception {
+
+        // first attempt to obtain a secret using a proper vault string with an existing key.
+        Optional<WeakReference<String>> optional = transcriber.getStringSecret("${vault.vault_key_1}");
+        Assert.assertTrue(optional.isPresent());
+        WeakReference<String> stringSecret = optional.get();
+        Assert.assertEquals("secret1", stringSecret.get());
+
+        // now attempt to obtain a secret using a proper vault string with a non-existing key.
+        optional = transcriber.getStringSecret("${vault.non_existing_key}");
+        Assert.assertFalse(optional.isPresent());
+
+        // finally try to obtain a secret using a string that is not a vault string. The string itself should be returned.
+        optional = transcriber.getStringSecret("non_vault_secret");
+        Assert.assertTrue(optional.isPresent());
+        stringSecret = optional.get();
+        Assert.assertEquals("non_vault_secret", stringSecret.get());
+
+        // check that a null expression results in an empty optional instance.
+        optional = transcriber.getStringSecret(null);
+        Assert.assertFalse(optional.isPresent());
+    }
+
+    @Test
+    public void testTranscriberWithNullProvider() {
+        VaultTranscriber transcriber = new DefaultVaultTranscriber(null);
+        // attempt to use the transcriber to obtain a secret using a valid vault expression (i.e. requires a proper provider)
+        try {
```

Comment:
Please do not use this way of testing expected exception. Use `@Rule ExpectedException` instead as illustrated e.g. here:
https://github.com/keycloak/keycloak/blob/master/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/admin/realm/RealmTest.java#L228

Reply:
Updated to use the @Rule